fastcrypto-zkp/src/conversions.rs

```diff
@@ -191,6 +195,171 @@ pub fn blst_g2_affine_to_bls_g2_affine(pt: &blst_p2_affine) -> BlsG2Affine {
+}
+
+struct EncodingFlags {
+    is_compressed: bool,
+    is_infinity: bool,
+    is_lexicographically_largest: bool,
+}
+
+impl From<&[u8]> for EncodingFlags {
+    fn from(bytes: &[u8]) -> Self {
+        Self {
+            is_compressed: ((bytes[0] >> 7) & 1) == 1,
+            is_infinity: (bytes[0] >> 6) & 1 == 1,
+            is_lexicographically_largest: (bytes[0] >> 5) & 1 == 1,
+        }
+    }
+}
+
+impl EncodingFlags {
+    fn encode_flags(&self, bytes: &mut [u8]) {
+        if self.is_compressed {
+            bytes[0] |= 1 << 7;
+        }
+
+        if self.is_infinity {
+            bytes[0] |= 1 << 6;
+        }
+
+        if self.is_compressed && !self.is_infinity && self.is_lexicographically_largest {
+            bytes[0] |= 1 << 5;
+        }
+    }
+}
+
+/// This deserializes an Arkworks G1Affine point from a Zcash point encoding.
+pub fn bls_g1_affine_from_zcash_bytes(bytes: &[u8; G1_COMPRESSED_SIZE]) -> Option<BlsG1Affine> {
+    // Obtain the three flags from the start of the byte sequence
+    let flags = EncodingFlags::from(&bytes[..]);
+
+    if !flags.is_compressed {
+        return None; // We only support compressed points
+    }
+
+    if flags.is_infinity {
+        return Some(BlsG1Affine::default());
+    }
+    // Attempt to obtain the x-coordinate
+    let x = {
+        let mut tmp = [0; G1_COMPRESSED_SIZE];
+        tmp.copy_from_slice(&bytes[0..48]);
+
+        // Mask away the flag bits
+        tmp[0] &= 0b0001_1111;
+
+        bls_fq_from_zcash_bytes(&tmp)?
+    };
+
+    BlsG1Affine::get_point_from_x(x, flags.is_lexicographically_largest)
+}
+
+/// This serializes an Arkworks G1Affine point into a Zcash point encoding.
+pub fn bls_g1_affine_to_zcash_bytes(p: &BlsG1Affine) -> [u8; G1_COMPRESSED_SIZE] {
+    let mut result = bls_fq_to_zcash_bytes(&p.x);
+    let encoding = EncodingFlags {
+        is_compressed: true,
+        is_infinity: p.infinity,
+        is_lexicographically_largest: p.y > -p.y,
+    };
+    encoding.encode_flags(&mut result[..]);
+    result
+}
+
+/// This deserializes an Arkworks G2Affine point from a Zcash point encoding.
+pub fn bls_g2_affine_from_zcash_bytes(bytes: &[u8; G2_COMPRESSED_SIZE]) -> Option<BlsG2Affine> {
+    // Obtain the three flags from the start of the byte sequence
+    let flags = EncodingFlags::from(&bytes[..]);
+
+    if !flags.is_compressed {
+        return None; // We only support compressed points
+    }
+
+    if flags.is_infinity {
+        return Some(BlsG2Affine::default());
+    }
+
+    // Attempt to obtain the x-coordinate
+    let xc1 = {
+        let mut tmp = [0; G1_COMPRESSED_SIZE];
+        tmp.copy_from_slice(&bytes[0..48]);
+
+        // Mask away the flag bits
+        tmp[0] &= 0b0001_1111;
+
+        bls_fq_from_zcash_bytes(&tmp)?
+    };
+    let xc0 = {
+        let mut tmp = [0; G1_COMPRESSED_SIZE];
+        tmp.copy_from_slice(&bytes[48..96]);
+
+        bls_fq_from_zcash_bytes(&tmp)?
+    };
+
+    let x = Fq2::new(xc0, xc1);
```

Comment:
what is `Fq2`?

Reply:
`Fq2` is the quadratic extension field which forms the base field over which G2 is defined.